File: 03-march/12-maximize-spanning-tree-stability-with-upgrades/solution.py
class DSU:
    def __init__(self, n: int) -> None:
        self.parent = list(range(n))
        self.size = [1] * n
        self.components = n

    def find(self, x: int) -> int:
        while self.parent[x] != x:
            self.parent[x] = self.parent[self.parent[x]]
            x = self.parent[x]
        return x

    def union(self, a: int, b: int) -> bool:
        ra = self.find(a)
        rb = self.find(b)
        if ra == rb:
            return False
        if self.size[ra] < self.size[rb]:
            ra, rb = rb, ra
        self.parent[rb] = ra
        self.size[ra] += self.size[rb]
        self.components -= 1
        return True


class Solution:
    def maxStability(self, n: int, edges: list[list[int]], k: int) -> int:
        must_edges = []
        optional_edges = []
        hi = 0

        for u, v, s, must in edges:
            if must == 1:
                must_edges.append((u, v, s))
                hi = max(hi, s)
            else:
                optional_edges.append((u, v, s))
                hi = max(hi, 2 * s)

        def feasible(x: int) -> bool:
            dsu = DSU(n)

            for u, v, s in must_edges:
                if s < x:
                    return False
                if not dsu.union(u, v):
                    return False

            for u, v, s in optional_edges:
                if s >= x:
                    dsu.union(u, v)

            upgrades = 0
            for u, v, s in optional_edges:
                if s < x <= 2 * s and dsu.union(u, v):
                    upgrades += 1
                    if upgrades > k:
                        return False

            return dsu.components == 1

        if not feasible(1):
            return -1

        lo = 1
        ans = 1
        while lo <= hi:
            mid = (lo + hi) // 2
            if feasible(mid):
                ans = mid
                lo = mid + 1
            else:
                hi = mid - 1

        return ans

File: 03-march/12-maximize-spanning-tree-stability-with-upgrades/solution.java
class Solution {
    private static class DSU {
        int[] parent;
        int[] size;
        int components;

        DSU(int n) {
            parent = new int[n];
            size = new int[n];
            components = n;
            for (int i = 0; i < n; i++) {
                parent[i] = i;
                size[i] = 1;
            }
        }

        int find(int x) {
            while (parent[x] != x) {
                parent[x] = parent[parent[x]];
                x = parent[x];
            }
            return x;
        }

        boolean union(int a, int b) {
            int ra = find(a);
            int rb = find(b);
            if (ra == rb) {
                return false;
            }
            if (size[ra] < size[rb]) {
                int tmp = ra;
                ra = rb;
                rb = tmp;
            }
            parent[rb] = ra;
            size[ra] += size[rb];
            components--;
            return true;
        }
    }

    public int maxStability(int n, int[][] edges, int k) {
        java.util.ArrayList<int[]> mustEdges = new java.util.ArrayList<>();
        java.util.ArrayList<int[]> optionalEdges = new java.util.ArrayList<>();
        int hi = 0;

        for (int[] e : edges) {
            int u = e[0], v = e[1], s = e[2], must = e[3];
            if (must == 1) {
                mustEdges.add(new int[]{u, v, s});
                hi = Math.max(hi, s);
            } else {
                optionalEdges.add(new int[]{u, v, s});
                hi = Math.max(hi, 2 * s);
            }
        }

        if (!feasible(1, n, k, mustEdges, optionalEdges)) {
            return -1;
        }

        int lo = 1;
        int ans = 1;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (feasible(mid, n, k, mustEdges, optionalEdges)) {
                ans = mid;
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }

        return ans;
    }

    private boolean feasible(int x, int n, int k, java.util.ArrayList<int[]> mustEdges,
                             java.util.ArrayList<int[]> optionalEdges) {
        DSU dsu = new DSU(n);

        for (int[] e : mustEdges) {
            int u = e[0], v = e[1], s = e[2];
            if (s < x) {
                return false;
            }
            if (!dsu.union(u, v)) {
                return false;
            }
        }

        for (int[] e : optionalEdges) {
            int u = e[0], v = e[1], s = e[2];
            if (s >= x) {
                dsu.union(u, v);
            }
        }

        int upgrades = 0;
        for (int[] e : optionalEdges) {
            int u = e[0], v = e[1], s = e[2];
            if (s < x && 2 * s >= x && dsu.union(u, v)) {
                upgrades++;
                if (upgrades > k) {
                    return false;
                }
            }
        }

        return dsu.components == 1;
    }
}
